<div class="sect1">
<h2 id="_description">Description</h2>
<div class="sectionbody">

</div>
</div>
<div class="sect1">
<h2 id="_why_is_this_an_issue">Why is this an issue?</h2>
<div class="sectionbody">
<div class="paragraph">
<p>When logging a message, the following requirements should be fulfilled:</p>
</div>
<div class="ulist">
<ul>
<li>
<p>The user can easily record logged data and retrieve the logs.</p>
</li>
<li>
<p>The format of all logged messages is uniform and provides context about the message creator. This improves the readability of the logs.</p>
</li>
<li>
<p>Sensitive data is logged securely.</p>
</li>
</ul>
</div>
<div class="paragraph">
<p>Simply printing messages to <code>System.out</code> or <code>System.err</code> does not fulfill these needs.
A dedicated logger should be used instead.</p>
</div>
<div class="sect2">
<h3 id="_noncompliant_code_example">Noncompliant code example</h3>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-java" data-lang="java">class PrintMessage {
  public void print() {
    System.out.println("My Message");  // Noncompliant, output directly to System.out without a logger
  }
}</code></pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_compliant_solution">Compliant solution</h3>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-java" data-lang="java">import java.util.logging.Logger;

class PrintMessage {

  Logger logger = Logger.getLogger(getClass().getName());

  public void print() {
    logger.info("My Message");  // Compliant, output via logger
  }
}</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_resources">Resources</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_documentation">Documentation</h3>
<div class="ulist">
<ul>
<li>
<p><a href="https://docs.oracle.com/javase/7/docs/api/java/util/logging/Logger.html">Java SE 7 API Specification: java.util.logging.Logger</a></p>
</li>
<li>
<p><a href="https://owasp.org/Top10/A09_2021-Security_Logging_and_Monitoring_Failures/">OWASP Top 10 2021 Category A9</a> - Security Logging and Monitoring Failures</p>
</li>
<li>
<p><a href="https://www.owasp.org/www-project-top-ten/2017/A3_2017-Sensitive_Data_Exposure">OWASP Top 10 2017 Category A3</a> - Sensitive Data Exposure</p>
</li>
<li>
<p><a href="https://wiki.sei.cmu.edu/confluence/x/nzdGBQ">CERT, ERR02-J.</a> - Prevent exceptions while logging data</p>
</li>
</ul>
</div>
<hr>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_implementation_specification">Implementation Specification</h2>
<div class="sectionbody">
<div class="paragraph">
<p>(visible only on this page)</p>
</div>
<div class="sect2">
<h3 id="_message">Message</h3>
<div class="paragraph">
<p>Replace this use of System.out or System.err with a logger.</p>
</div>
<hr>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_comments_and_links">Comments And Links</h2>
<div class="sectionbody">
<div class="paragraph">
<p>(visible only on this page)</p>
</div>
<div class="sect2">
<h3 id="_deprecates_s2228">deprecates: <a data-rspec-id="S2228" class="rspec-auto-link">S2228</a></h3>

</div>
<div class="sect2">
<h3 id="_on_8_may_2013_095007_freddy_mallet_wrote">on 8 May 2013, 09:50:07 Freddy Mallet wrote:</h3>
<div class="paragraph">
<p>{panel:title=Java Title}</p>
</div>
<div class="paragraph">
<p>System.(out/err) should not be used to log anything</p>
</div>
<div class="paragraph">
<p>{panel}</p>
</div>
<div class="paragraph">
<p>{panel:title=C&#43;&#43; Title}</p>
</div>
<div class="paragraph">
<p>cout and cerr should not be used to log anything</p>
</div>
<div class="paragraph">
<p>{panel}</p>
</div>
</div>
<div class="sect2">
<h3 id="_on_6_aug_2013_140558_dinesh_bolkensteyn_wrote">on 6 Aug 2013, 14:05:58 Dinesh Bolkensteyn wrote:</h3>
<div class="paragraph">
<p>Implemented by https://jira.codehaus.org/browse/SONARJAVA-234</p>
</div>
</div>
</div>
</div>