<div class="sect1">
<h2 id="_description">Description</h2>
<div class="sectionbody">

</div>
</div>
<div class="sect1">
<h2 id="_why_is_this_an_issue">Why is this an issue?</h2>
<div class="sectionbody">
<div class="paragraph">
<p>Leaving unused props in a React component can make the code harder to understand and maintain. Other developers may wonder why certain props are passed to a component if they are not used. Unused props can also increase the size of the component&#8217;s memory footprint and impact performance. This is especially true if the unused props are large objects or arrays. Furthermore, if a prop is unused, it may indicate that the developer did not complete the implementation as he intended initially or made a mistake while writing the component.</p>
</div>
<div class="paragraph">
<p>To avoid these issues, you should remove any unused props from React components. This helps keep the codebase clean, improves performance, and enhances code readability.</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_how_to_fix_it_in_proptypes">How to fix it in PropTypes</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_code_examples">Code examples</h3>
<div class="sect3">
<h4 id="_noncompliant_code_example">Noncompliant code example</h4>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-text" data-lang="text">import PropTypes from 'prop-types';
import React from 'react';

class Hello extends React.Component {
  render() {
    return (
      &lt;h1&gt;Hello&lt;/h1&gt;
    );
  }
}

Hello.propTypes = {
  name: PropTypes.string
};</code></pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_compliant_solution">Compliant solution</h4>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-text" data-lang="text">import PropTypes from 'prop-types';
import React from 'react';

class Hello extends React.Component {
  render() {
    return (
      &lt;h1&gt;Hello {this.props.name}&lt;/h1&gt;
    );
  }
}

Hello.propTypes = {
  name: PropTypes.string
};</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_how_to_fix_it_in_typescript">How to fix it in TypeScript</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_code_examples_2">Code examples</h3>
<div class="sect3">
<h4 id="_noncompliant_code_example_2">Noncompliant code example</h4>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-text" data-lang="text">import React from 'react';

type Props = {
  name: string;
}

class Hello extends React.Component&lt;Props&gt; {
  render() {
    return &lt;div&gt;Hello&lt;/div&gt;;
  }
}</code></pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_compliant_solution_2">Compliant solution</h4>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-text" data-lang="text">import React from 'react';

type Props = {
  name: string;
};

class Hello extends React.Component&lt;Props&gt; {
  render() {
    return &lt;div&gt;Hello {this.props.name}&lt;/div&gt;;
  }
}</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_resources">Resources</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_documentation">Documentation</h3>
<div class="ulist">
<ul>
<li>
<p><a href="https://react.dev/learn/passing-props-to-a-component">React - Passing Props to a Component</a></p>
</li>
<li>
<p><a href="https://react.dev/reference/react/Component#static-proptypes">React - static propTypes</a></p>
</li>
<li>
<p><a href="https://react.dev/learn/typescript#typescript-with-react-components">React - TypeScript with React Components</a></p>
</li>
</ul>
</div>
</div>
</div>
</div>