<div class="sect1">
<h2 id="_description">Description</h2>
<div class="sectionbody">
<div class="paragraph">
<p>This rule raises an issue when legacy <code>numpy.random.RandomState</code> is used.</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_why_is_this_an_issue">Why is this an issue?</h2>
<div class="sectionbody">
<div class="paragraph">
<p>Using a predictable seed is a common best practice when using NumPy to create reproducible results. To that end, using <code>np.random.seed(number)</code> to set the seed of the global <code>numpy.random.RandomState</code> has been the privileged solution for a long time.</p>
</div>
<div class="paragraph">
<p><code>numpy.random.RandomState</code> and its associated methods rely on a global state, which may be problematic when threads or other forms of concurrency are involved. The global state may be altered and the global seed may be reset at various points in the program (for instance, through an imported package or script), which would lead to irreproducible results.</p>
</div>
<div class="paragraph">
<p>Instead, the preferred best practice to generate reproducible pseudorandom numbers is to instantiate a <code>numpy.random.Generator</code> object with a seed and reuse it in different parts of the code. This avoids the reliance on a global state. Whenever a new seed is needed, a new generator may be created instead of mutating a global state.</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_how_to_fix_it">How to fix it</h2>
<div class="sectionbody">
<div class="paragraph">
<p>To fix this issue, replace usages of <code>numpy.random.RandomState</code> to <code>numpy.random.Generator</code>.</p>
</div>
<div class="sect2">
<h3 id="_code_examples">Code examples</h3>
<div class="sect3">
<h4 id="_noncompliant_code_example">Noncompliant code example</h4>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-python" data-lang="python">import numpy as np
def foo():
    np.random.seed(42)
    x = np.random.randn()  # Noncompliant: this relies on numpy.random.RandomState, which is deprecated</code></pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_compliant_solution">Compliant solution</h4>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-python" data-lang="python">import numpy as np
def foo():
    generator = np.random.default_rng(42)
    x = generator.standard_normal()</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_resources">Resources</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_documentation">Documentation</h3>
<div class="ulist">
<ul>
<li>
<p>NumPy Documentation - <a href="https://numpy.org/doc/stable/reference/random/generator.html#random-generator">Random Generator</a></p>
</li>
<li>
<p>NumPy Documentation - <a href="https://numpy.org/doc/stable/reference/random/legacy.html#legacy-random-generation">Legacy Random Generation</a></p>
</li>
<li>
<p>NumPy Documentation - <a href="https://numpy.org/neps/nep-0019-rng-policy.html">NEP19 RNG Policy</a></p>
</li>
</ul>
</div>
</div>
</div>
</div>